Add semi-automatic tests script

sculpt0r opened the pull request:
I create a script that uses [GH CLI](https://cli.github.com/manual/) & [GH API](https://docs.github.com/en/rest)

**Requirements**: pre-prepared repository with at least one commit.

**Workflow**:

1.  Switch to your working branch at `ckeditor4-workflows-common`, 
2.  Run `./test.sh`
    1.  Reset a master branch to a given commit. It could be any commit that leaves the master branch in the state needed for testing.
    2.  Delete all running (not queued!) actions
    3.  Commit all files inside `workflows` directory and optionally `workflows-settings.json` to target testing repository.
    4.  Gather and dispatch every action. Please notice, that actions have to be `dispatchable`. 

This script could be used in the current master branch state. If we provide more `.yml`'s that require e.g. push an event to be triggered - we will have to upgrade this script.

The script needs to be modified inside, at least with the testing repository and default init commit hash. So if we need to test workflow in a certain repo state - we can provide such a commit and the master branch will be always reset to this point.

The last point is to unfortunately manually check logs for each workflow. However, it's still faster to set up the initial environment & upload workflows files.

Closes #14

sculpt0r committed:
WIP: testing workflows script

sculpt0r committed:
test wf script fetching running actions

sculpt0r committed:
Fetch master commit sha

sculpt0r committed:
Create branch from master

sculpt0r committed:
Reorder commands

sculpt0r committed:
Clean up test order

sculpt0r committed:
Iterate over received actions

sculpt0r committed:
Iterate over workflow files (todo URL)

sculpt0r committed:
Make proper endpoint URL variable

sculpt0r committed:
Reset repo master branch to hardcoded commit

sculpt0r committed:
Display steps, order actions

sculpt0r committed:
Add delay before gather actions & dispatch all

sculpt0r committed:
Parse GH output

f1ames commented:
> **Workflow**:
> 
> Switch to your working branch at `ckeditor4-workflows-common`,
> 
> Run `./test.sh`
> 
> 1.  Reset a master branch to a given commit. It could be any commit that leaves the master branch in the state needed for testing.
> 2.  Delete all running (not queued!) actions
> 3.  Commit all files inside `workflows` directory and optionally `workflows-settings.json` to target testing repository.
> 4.  Gather and dispatch every action. Please notice, that actions have to be `dispatchable`.

That's a good start and you have already proven that it makes testing easier in #12 :+1:, but I think it will be good to polish testing procedure slightly. There are few problematic assumptions here:

*   Resetting branch to specific (hardcoded) commit.
*   Dispatching all workflows at once (what if conflicting things happens?).
*   Not checking workflow results at all.

We should go with the simplest procedure possible but at the same time make sure it's stable, workflows will not be conflicting one another and new tests could be added easily. So from my perspective:

1.  Instead of resetting test branch, maybe we can reset its contents by simply pushing new commit there? That way the commit can have proper description (like "Test workflow something, something") and then it will be clearly visible at what branch state each workflow was launched.
2.  Workflows should be launched in a sequence so we are sure there are now two workflows executing at the same time and interfering each other.
3.  Each workflow has specific outcome. But the first thing we can check automatically is if workflow was successful (so exit code 0). I hope GH API allows that (looks like [it could be done](https://docs.github.com/en/rest/reference/actions#list-workflow-runs)). If as a result workflow creates a branch or PR we can check if it was created. Or if new commit was pushed to specific branch. **I think as a first step we could focus on checking if workflow was successful only and then see how it works and what can be improved**.

As for more details on the above:

## Resetting branch contents

It will be good to do this without force pushing so we have entire history preserved for easier workflows "debugging". I think it could be done like:

```
git clone test-repo
cd test-repo
rm -rf *
// move files required for specific test
git add .
git commit -m "Test..."
git push origin branch-name
```

This way for each test we will have files required but at the same time the history is preserved and can be checked later. TBH this looks like a `beforeEach` step so probably should be run before each test.

## Launching workflows in a sequence and checking its result

I would imagine each workflow run as a separate test in a test suite - so `beforeEach` is run then a single test (dispatching workflow and waiting for it to end) and then go to next test/workflow. This prevents workflows from interfering each other and makes it possible to run single workflow more than once with different configuration.

And naturally this requires checking if and how each workflow ended. As mentioned above - **I think as a first step we could focus on checking if workflow was successful only and then see how it works and what can be improved**.

## Further issues

For now I see at least two issue with the current and proposed approach:

*   Since we use single repo for tests it might be problematic if two people would run tests at the same time (or when we add CI here). But I guess we will think about it later since it's not an issue for now.
*   Clean up phase - some workflows create branches or PRs, which means second run will behave different due to those leftovers. For the beginning we could cleanup manually (as we will be checking runs results manually too), but later it might be good to think about clean-up step.

---

**That's my suggestions only, so feel free to propose different approach or any improvements or simplifications.** And well, I didn't plan to create such a "wall of text", but hopefully it's descriptive enough. If not, we could talk F2F to clarify any details.

f1ames reviewed (changes requested):
Please see https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-781181435.

sculpt0r committed:
WIP: continuosly action state verification

sculpt0r committed:
Add JS testing approach

sculpt0r commented:
Rebase

sculpt0r committed:
WIP: add before-each script

sculpt0r committed:
WIP: not working git add

sculpt0r committed:
Update / upload file at git repo

sculpt0r committed:
Cleanup before-each

sculpt0r committed:
Fix typo: additional curly bracket

sculpt0r committed:
Test full circle (with single file upload)

sculpt0r committed:
Multiple files upload

sculpt0r committed:
Add editor config

sculpt0r committed:
Clean up code, removed unused code, fix intendation

sculpt0r committed:
Remove debugger instruction

sculpt0r committed:
Get rid of octokit deps from test logic

sculpt0r committed:
Fix code style & add chalk to stylize output

sculpt0r committed:
Switch to async/await & run multiple tests

sculpt0r committed:
Remove unused script

sculpt0r committed:
Add input option for dispatched workflow

sculpt0r commented:
@f1ames Ready for another review

I rewrite API usage to JS.

There is `.env` file required with at least `AUTH_KEY`, `OWNER`, `REPO` to work correctly.  
  
There are two tests for setup workflow & dependency updater.  
Also, after a run `node index.js` there are results of both tests in the console.  
  
I decide to push files to the repository via GH API, rather than making local operations on the real directory. Currently, each file is committed as a separate commit.

f1ames commented on index.js:
```diff
@@ -0,0 +1,141 @@
+	GitHubClient = require ( './github-client' ),
+	SendFiles = require( './before-each' );
+
+dotenv.config();
```
I think it will be good to have some env variables validation so one can clearly see what's missing.

f1ames replied:
Also short section on testing added to README file could be helpful.

f1ames commented on before-each.js:
```diff
@@ -0,0 +1,71 @@
+const fs = require( 'fs' ),
+	dotenv = require( 'dotenv' ),
+	GitHubClient = require ( './github-client' );
```
I think we don't have code style guide regarding this, but in other projects (e.g. OpenXML Converter) each require is separate `const`.

f1ames commented on before-each.js:
```diff
@@ -0,0 +1,71 @@
+
+dotenv.config();
+
+async function CommitFile( sha, content, filePath, branch ) {
```
I'm not sure if there was any particular reason to go with function names starting with uppercase?

sculpt0r replied:
It wasn't... :cry:  I correct this :)

f1ames commented on before-each.js:
```diff
@@ -0,0 +1,71 @@
+function SendFile(branch, sourceFilePath, destinationFilePath) {
+	return new Promise( ( resolve, reject ) => {
+
+	const newContent = fs.readFileSync( sourceFilePath, {
```
Somehow `async` is mixed with promises here(I mean it's just the syntax since both use promises underneath). From the outside perspective both can be used with `await` so it's only internal inconsistency.

Now I wonder, `SendFile` and `SendFiles` returns a promise and is used with `await` in the code - shouldn't it have `async` too for consistency? :thinking:

f1ames commented on before-each.js:
```diff
@@ -0,0 +1,71 @@
+		encoding: 'base64'
+	} );
+
+	GetFile( destinationFilePath, branch )
```
And here you are using `GetFile` with `then` instead of `await` :thinking:

f1ames commented on before-each.js:
```diff
@@ -0,0 +1,71 @@
+const fs = require( 'fs' ),
```
I guess those are some "before each helpers"? Not a code which is executed as is before each test? Maybe we could simply have some helpers file with all the functions used to run tests (e.g. the ones in `index.js` are also used to setup/teardown/verify tests)?

f1ames replied:
Or you can try to split it contextually - like helpers for file handling, for workflows handling and so on?

f1ames commented on index.js:
```diff
@@ -0,0 +1,141 @@
+'use strict';
```
This file is kind of "test runner" so maybe we could rename it? OTOH `index.js` in `tests` directory sounds like a valid entry point :thinking:

f1ames commented on index.js:
```diff
@@ -0,0 +1,141 @@
+
+dotenv.config();
+const allTestsResults = [];
+const tests = [
```
Maybe we could extract this config to a separate file?

f1ames reviewed (changes requested):
Starts to looking good :+1: I really like colored output :rainbow: :+1: 

Some things to polish still:

1. `npm test` script could be updated so it runs test command.
2. Maybe all test related file could be moved to `tests` directory?
3. Have you checked if it will be doable to remove all but `main`/`master` branches on the beginning of a test run via GH API? I wonder if we should just do global clean up or maybe each test should clean after itself (the first one sounds much easier TBH).

As for "Next check in" message in the output maybe it can be merged with a previous line? It's not crucial information, OTHO it's good that it's there so one can see what's going on. WDYT?

![image](https://user-images.githubusercontent.com/1061942/113708401-76cebb80-96e1-11eb-91fa-817eb9c968c2.png)

Apart from that please see review comments.

sculpt0r committed:
Lowercase function names, paritialy remove new Promise

sculpt0r committed:
Merge console logs

sculpt0r committed:
Add env variables veryfication

sculpt0r committed:
Reorganize test files

sculpt0r committed:
Add tests readme

sculpt0r committed:
Update README, testing section

sculpt0r committed:
Move status logging to test runner

sculpt0r committed:
Move test cfg to tests dir. Code style fixes

sculpt0r commented:
@f1ames

> Have you checked if it will be doable to remove all but main/master branches on the beginning of a test run via GH API? I wonder if we should just do global clean up or maybe each test should clean after itself (the first one sounds much easier TBH).

I have to talk to you before review ;p I'm not sure what do you mean by clean after itself ;p

Since you gave me a lot of choices, here is a list of modifications:
- Command `npm test` now runs tests (`node tests/index.js`).
- Test-related files moved to `tests` directory.
- Worflow status update (in console) merged with `next check` time information.
- Added `env` variables validation with proper output & process exit.
- Added section about testing in README.
- Each `require` has own `const` declaration.
- Functions names start with a lowercase letter.
- I removed explicit `new Promise` in few places, however - when I need to call `resolve` - I preserved it
- Reorganize files structure. Now we have a test-runner, GitHub helpers for workflows & commits, GitHub client, and config. The main entry point (`index`) contains tests definition and runs them.

sculpt0r commented:
As we talked with @f1ames 
We leave commits created during testing to have an exact view on action & commit related to it. In case you forgot which changes you made and how the results look like.

In the end - all of this is made (should be! ;p) on testing repo.

sculpt0r commented:
Ready for review.

f1ames commented on tests/config.js:
```diff
@@ -0,0 +1,28 @@
+
+			if ( !anyMissingVariable ) {
+				anyMissingVariable = true;
+			}
```
```suggestion
			anyMissingVariable = true;
```

:point\_up: No need for `if` here I guess. Even if it's already `true` it will be just set once again. Alternatively you can go with:

```javascript
let allVariablesSet = true;
...
allVariablesSet = allVariablesSet && !process.env[ variable ];
```

but in this case it seems not necessary.

f1ames committed:
Improve README file.

f1ames committed:
Update package-lock.json file.

f1ames commented on tests/_helpers/github-files.js:
```diff
@@ -0,0 +1,53 @@
+async function sendFiles( branch, files ) {
+	const nextFile = files.shift();
+
+	if( nextFile ) {
+		await sendFile( branch, nextFile.src, nextFile.dest );
+		return sendFiles( branch, files );
+	}
```
What will it return if there is no `nextFile`?

f1ames replied:
Have you considered using `for...of` loop instead of recursion here? Like:

```javascript
for ( const file of files ) {
  await sendFile( ... )
}
```

?

f1ames commented on tests/_helpers/github-files.js:
```diff
@@ -0,0 +1,53 @@
+	} );
+
+	const file = await getFile( destinationFilePath, branch );
+	await commitFile( file.data.sha, newContent, destinationFilePath, branch );
```
Maybe `sendFile` could also return a result of `commitFile`? Or anything which will tell that commiting filie was successful or failed?

f1ames commented on tests/_helpers/github-workflows.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+async function getWorkflowRun( workflowId ) {
+	const result = await GitHubClient.request(
+		'GET',
+		'/repos/{owner}/{repo}/actions/runs/{run_id}',
```
In other places like [https://github.com/ckeditor/ckeditor4-workflows-common/pull/15/files#diff-6f6766ef88da83cef9c00870cc460545c8ce3125106eb5dd135027274cf06b0cR23,](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15/files#diff-6f6766ef88da83cef9c00870cc460545c8ce3125106eb5dd135027274cf06b0cR23,) you use single line for this, and here is split into multiple lines - it will be good to keep it consistent across entire code base.

sculpt0r replied:
Seems to be a little too long for me at that time. Inlined :)

f1ames commented on tests/_helpers/github-workflows.js:
```diff
@@ -0,0 +1,70 @@
+	return result;
+}
+
+async function getWorkflowRun( workflowId ) {
```
Mayeb `getWorkflowRunResult`?

f1ames commented on tests/_helpers/github-workflows.js:
```diff
@@ -0,0 +1,70 @@
+	} );
+}
+
+async function getRunningActions() {
```
I guess it will be good to stick to one name so `Workflow` vs `Action`. From what I see `Workflow` should be used :thinking:

f1ames commented on tests/_helpers/github-workflows.js:
```diff
@@ -0,0 +1,70 @@
+const GitHubClient = require( './github-client' );
+
+function verifyWorkflowStatus( workflowObject, reportCurrentStatus, waitingTime ) {
+	return new Promise( ( resolve, reject ) => {
+		if( workflowObject.status === 'completed' ) {
+			resolve( workflowObject );
+			return;
+		}
+		waitingTime = waitingTime || 1000;
+
+		reportCurrentStatus( workflowObject, waitingTime );
+
+		setTimeout( async () => {
+			const workflow = await getWorkflowRun( workflowObject.id );
+			// Give some time between rechecks
+			const result = await verifyWorkflowStatus( workflow.data, reportCurrentStatus, waitingTime + 3500 );
+			resolve( result );
+		}, waitingTime );
+	} );
```
WDYT about making it `async` function only? It will only require wrapping `setTimeout` into a promise (like [mentioned here](https://stackoverflow.com/a/33292942/646871)) and then you can get rid of `new Promise` wrapper and go with `await` only.

f1ames commented on tests/_helpers/test-runner.js:
```diff
@@ -0,0 +1,42 @@
+const { dispatchWorkflow, verifyWorkflowStatus, getRunningActions } = require( './github-workflows' );
+
+async function runTests( tests ) {
+	const testCase = tests.shift();
+
+	if( testCase ) {
+		console.log( '\n *** Running test for: ' + chalk.blue( testCase.name ) + ' ***' );
+		await runTest( testCase );
+		return runTests( tests );
+	}
```
Again, maybe `for...of` loop instead of recursion?

f1ames commented on tests/_helpers/test-runner.js:
```diff
@@ -0,0 +1,42 @@
+}
+
+async function runTest( testCase ) {
+	return new Promise( async ( resolve, reject ) => {
```
I have the same proposal for getting rid of `new Promise` as for the `verifyWorkflowStatus()`. WDYT?

f1ames commented on tests/index.js:
```diff
@@ -0,0 +1,33 @@
+require( './config' )();
+const { runTests } = require( './_helpers/test-runner' );
+
+const tests = [
```
Ok, so it will be good to have the ultimate test for our runner. Now we have only two tests here (since we have two testable workflows only). But both workflows can have different internal flows.

So `setup-workflows.yml` may push directly to master or create a PR. Same for `update-deps.yml`, it can have `targetBranch` changed via config or can be run on repo withut `package.json` file (btw. it is missing in this test now).

If we are able to create reasonable working tests covering all the above, we are good to go. WDYT?

f1ames reviewed (changes requested):
Ok, some things to polish still but looking good already :+1: 

Maybe we could have dir structure like:

- tests
  - helpers/
    - github/
      - client.js
      - files.js
      - ...
  - fixtures/
    - workflows-config.js
  - index.js
  - config.js

I know it may seem a little over complicated but OTOH it's easier to see files purpose for each dir.

Please also see review comments.

sculpt0r committed:
Correct code style

sculpt0r committed:
Rename functions, verify result of commit files

sculpt0r committed:
Get rid of 'new Promise', add errors detection

sculpt0r committed:
REorganize directory structure

sculpt0r committed:
[WIP] Rework test fixtures

sculpt0r committed:
Working test runner

sculpt0r committed:
Fixture directery iteration refactor

sculpt0r committed:
Add last empty line in .gitignore

sculpt0r committed:
Fixtures gathering refactor, rework README

sculpt0r committed:
Return & display error in GH client

sculpt0r committed:
Correct fixtures condition, correct comment

sculpt0r commented:
@f1ames - ready for review
I made some ground rework, but also correct all of your comments. However, I have one thing to discuss with you ;)
- I get rid of all `new Promise`. Indeed. code looks much better now. I have to write more async code ;p;p
- Change recursion to `for...of` loops.
- Made some comments corrections.

I made a ground rework about preparing fixtures to test. Now they are separated into folders. Also added one config field, which is `workflow` and it has to match workflow `yml` config file. This file will be automatically added to the files list. Also, all additional files can be placed in the fixture directory. In the config file, the path to those files has to be related just to this fixture directory. Because of that, I correct some entries in README.

Also, I added three test cases, which are mentioned [here](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#discussion_r613046573). Those are **not** all of them, but:
- this kind of preparation (like remove `package.json` from repo) to a fully automated test requires more custom code to be written in each test case.
- As I remember current approach has to cover something like auto send required files and run a selected workflow. If you need any special environment in your repository, you need to create it manually and change `REPO` variable in `.env` to point to that repo. I know it's not ideal - but as you said somewhere before - we need to start somewhere.

Probably the next step can introduce custom GH API calls in fixtures. So we have to extend the test fixture to not only contain some configuration but also to be able to execute some code. This could be another story, based on our previous talks.

sculpt0r committed:
Fix code style

f1ames commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+node_modules
+.env
+
```
A wild empty line appears! :wink: :stuck\_out\_tongue:

sculpt0r replied:
As we talked, it's better to have it added now :)

f1ames commented on README.md:
```diff
@@ -111,3 +111,45 @@ Workflow responsible for updating NPM dependencies. It is run once a week (at 05
+
+### Adding tests
+
+New tests should be added in `tests/index.js` file. Each test has similar structure:
```
Since test file structure was changed we need to slightly adjust README too.

sculpt0r replied:
:see_no_evil: I somehow missed that sentence.

f1ames commented on tests/helpers/github/files.js:
```diff
@@ -0,0 +1,64 @@
+		encoding: 'base64'
+	} );
+
+	try{
```
```suggestion
	try {
```

:wink:

f1ames reviewed (changes requested):
> * I get rid of all `new Promise`. Indeed. code looks much better now. I have to write more async code ;p;p
> * Change recursion to `for...of` loops.
> * Made some comments corrections.

:+1: 

> I made a ground rework about preparing fixtures to test. Now they are separated into folders. Also added one config field, which is `workflow` and it has to match workflow `yml` config file. This file will be automatically added to the files list. Also, all additional files can be placed in the fixture directory. In the config file, the path to those files has to be related just to this fixture directory. Because of that, I correct some entries in README.

Looks better now :+1: However, I'm not sure if it's not too fragmented... Each single test is a separate dir now, with separate `index.js` and `workflows-config.json` file, which makes adding new tests cumbersome and you can't group tests into suits.
Apart from that, I see the config still has `fileLists` listing config file too which seems redundant:

```js
	filesList: [
		{
			src: 'workflows-config.json',
			dest: '.github/workflows-config.json'
		},
	]
```

What I meant in my previous https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#pullrequestreview-635302962, was to have `fixtures/` directory for additional files necessary for test runs, but not sure if it was clear enough, sorry :see_no_evil: 

So I would be for simplifying it slightly:

1. Group tests for single workflow in a single file.
2. Inline config into test declaration.
3. If any extra files are needed, put it into `assets/` dir.

And so:

* tests/
  * helpers/... (stay as is)
  * fixtures/
    * setup-workflows.js
    * update-deps.js
    * ...
  * assets/
    * testname-package.json
    * ...
  * index.js
  * config.js

And `fixtures/setup-workflows.js` could look like:

```js
module.exports = [ {
	name: 'direct push',
	workflow: 'setup-workflows.yml',
	branch: 'master'
}, {
	name: 'as PR',
	workflow: 'setup-workflows.yml',
	branch: 'master',
        config: {
        	setupWorkflows: {
			pushAsPullRequest: true
		}
        },
	filesList: []

} ];
```

Theoretically `workflow` is kind of redundat but I will leave it (and use it) instead of relaying on file name (yeah, kind of against "convention over configuration" paradigm but...).

WDYT?

---

> Also, I added three test cases, which are mentioned [here](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#discussion_r613046573). Those are **not** all of them, but:
> 
> * this kind of preparation (like remove `package.json` from repo) to a fully automated test requires more custom code to be written in each test case.
> * As I remember current approach has to cover something like auto send required files and run a selected workflow. If you need any special environment in your repository, you need to create it manually and change `REPO` variable in `.env` to point to that repo. I know it's not ideal - but as you said somewhere before - we need to start somewhere.
> 
> Probably the next step can introduce custom GH API calls in fixtures. So we have to extend the test fixture to not only contain some configuration but also to be able to execute some code. This could be another story, based on our previous talks.

I guess what we have now should be enough for the first version and then we can adjust according to what we really need. So 3 test cases are good for the start. As you mentioned it would require some preparation and clean-up step for each test, but then we would also need to save and verify result of each workflow run before it gets cleaned up.

sculpt0r commented:
> What I meant in my previous #15 (review), was to have fixtures/ directory for additional files necessary for test runs, but not sure if it was clear enough, sorry see_no_evil

> So I would be for simplifying it slightly:

> Group tests for single workflow in a single file.
Inline config into test declaration.
If any extra files are needed, put it into assets/ dir.

I want to avoid 'all additional files' in one directory. But since there is not much of them. We can make it happen.

Inlining config into test declaration requires some additional changes, however, it looks more consistent and easier to maintain. So I'll rewrite it.

sculpt0r committed:
Relace chunked directory based fixtures with files

sculpt0r committed:
Add missing space after `try`

sculpt0r committed:
Improve test fixtures, pass config as request

sculpt0r committed:
Parse config as input for workflow

sculpt0r committed:
Fix path to assets

sculpt0r committed:
Fill sample package.json for update teps workflow

sculpt0r committed:
Add .nvmrc (14.16.1)

sculpt0r committed:
Accept 201 as valid request result

sculpt0r committed:
Update README

sculpt0r commented:
@f1ames - ready for review
- replaced fixture directories with one file per workflow
- config is now passed as part of workflow dispatch request, rather than a configuration file
- additional files are located in `assets` directory
- Cover case when we try to put a new file to the repository (so we haven't SHA for it)
- Add nvmrc to use node -v 14 (I used flatMap which is not supported in 10)
- Updated README

There is a request error visible if we try to get SHA for a non-existing file, but everything is working fine. This case is covered. However, I don't want to suppress errors just in case...

f1ames committed:
Polish tests README file.

f1ames committed:
Add new test case for setup-worfklows.

f1ames committed:
Rename filesList to fileList.

f1ames commented on tests/assets/deps-package.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Well, there are no `devDeps` or `deps` here, so the test will not check if any dependencies were correctly updated. I think default test case should cover that?

sculpt0r replied:
> When you run tests against empty repo it throws an error (but then proceeds anyway)

> From what I see the error is because there is no setup-workflows.yml file there. Why it checks it and reports this as an error? Since tests pushes this file I would assume it usually won't be there thinking

I mention that partially here: https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-824042833

GH Client throws an exception for `404` (which in this case is not an error, but useful information). Because there is no file on the repo, we don't want to extract SHA from that old file. We need to make this request - even if the URL is invalid for a non-existing file.

Previously I decided to `try...catch` any error right in the `client.js` module - to not spread them on the entire application. Request results are still returned, but also logged at the screen just in case. A similar 404 error could mean that someone provides the wrong GH rest API URL. But you can always verify your results via status. I don't want to `try...catch` each call in the application. So this log could be removed. Verifying status seems to be less code mess than wrapping everything in `try...catch` :thinking: WDYT?

f1ames reviewed (changes requested):
I would say it works quite nice when it comes to tests experience, and handling/dispatching/checking workflows :+1: 

Looks like a solid base for proper testing. I mention it as a base because we still have some work to cover to test different workflows scenarios, e.g. `setup-worfklows.yml` will push all workflow files in the first tests so then any other tests would skip this step. Also it requires some cleanup before running tests but that's what we agreed on :+1: 

There are two minor issues left, so after that it will be ready to go :fire: And we can then work on covering all the workflows in a separate issue (could you extract one @sculpt0r)?

---

When you run tests against empty repo it throws an error (but then proceeds anyway):

![image](https://user-images.githubusercontent.com/1061942/115712542-3162ed00-a375-11eb-972c-65dfe924d410.png)

From what I see the error is because there is no `setup-workflows.yml` file there. Why it checks it and reports this as an error? Since tests pushes this file I would assume it usually won't be there :thinking:

sculpt0r commented:
> When you run tests against empty repo it throws an error (but then proceeds anyway)

> From what I see the error is because there is no setup-workflows.yml file there. Why it checks it and reports this as an error? Since tests pushes this file I would assume it usually won't be there thinking

I mention that partially here: https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-824042833

GH Client throws an exception for `404` (which in this case is not an error, but useful information). Because there is no file on the repo, we don't want to extract SHA from that old file. We need to make this request - even if the URL is invalid for a non-existing file.

Previously I decided to `try...catch` any error right in the `client.js` module - to not spread them on the entire application. Request results are still returned, but also logged at the screen just in case. A similar 404 error could mean that someone provides the wrong GH rest API URL. But you can always verify your results via status. I don't want to `try...catch` each call in the application. So this log could be removed. Verifying status seems to be less code mess than wrapping everything in `try...catch` :thinking: WDYT?

sculpt0r committed:
Remove error logging from the screen

sculpt0r committed:
Add logs woth links to testing workflow & repo

sculpt0r commented:
Issues to extract:
- run test in two ways: `npm run test` and `npm run test fixtureName testCaseIndex`. With this additional run option, we could run a specific test case from the selected fixture. Much faster if we want to test just a single case. - https://github.com/ckeditor/ckeditor4-workflows-common/issues/24
- automatic clean up testing repo before each test case - https://github.com/ckeditor/ckeditor4-workflows-common/issues/23
- cover test cases for setup-workflows.yml - https://github.com/ckeditor/ckeditor4-workflows-common/issues/20
- cover test cases for update-deps.yml - https://github.com/ckeditor/ckeditor4-workflows-common/issues/21
- cover test cases for stalebot.yml (actually it looks doable - to test it on a correctly prepared repo) - https://github.com/ckeditor/ckeditor4-workflows-common/issues/22

sculpt0r committed:
Add fake deps and devDeps to testing package

sculpt0r commented:
@f1ames - ready for review
- extracted all cases. Full list is [here](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-825483434)
- removed error log, based on [this explanation](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-825452613)
- Add some fake and old deps and devDeps to `deps-package.json`
- Additionally, I put direct links to each runned workflow - it will be easier to check out details. Also add link to tested repository at the end of results.

f1ames commented:
Rebased onto latest `master`.

f1ames committed:
Add more information about test repo to readme file.

f1ames reviewed (approved):
LGTM :+1: :tada:

Let's move from there with the follow-ups - we will cover them when necessary.

---

> *   extracted all cases. Full list is [here](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-825483434)

Good job on extracting all those follow-ups :+1: 

> *   removed error log, based on [this explanation](https://github.com/ckeditor/ckeditor4-workflows-common/pull/15#issuecomment-825452613)

:+1: 

> *   Add some fake and old deps and devDeps to `deps-package.json`

:+1: 

> *   Additionally, I put direct links to each runned workflow - it will be easier to check out details. Also add link to tested repository at the end of results.

Nice touch :star\_struck:
